boot/systems.go

```diff
@@ -20,9 +20,127 @@
 package boot
 
 import (
+	"fmt"
+
 	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/bootloader"
 )
 
+type errInconsistentRecoverySystemState struct {
+	why string
+}
+
+func (e *errInconsistentRecoverySystemState) Error() string { return e.why }
+func IsInconsystemRecoverySystemState(err error) bool {
+	_, ok := err.(*errInconsistentRecoverySystemState)
+	return ok
+}
+
+// InitramfsIsTryingRecoverySystem typically called while in initramfs of
+// recovery mode system, checks whether the boot variables indicate that the
+// given recovery system is only being tried. When the state of boot variables
+// is inconsistent, eg. status indicates that a recovery system is to be tried,
+// but the label is unset, a specific error which can be tested with
+// InitramfsIsTryingRecoverySystem() is returned.
+func InitramfsIsTryingRecoverySystem(currentSystemLabel string) (bool, error) {
+	opts := &bootloader.Options{
+		// setup the recovery bootloader
+		Role: bootloader.RoleRecovery,
+	}
+	bl, err := bootloader.Find(InitramfsUbuntuSeedDir, opts)
+	if err != nil {
+		return false, err
+	}
+
+	vars, err := bl.GetBootVars("try_recovery_system", "recovery_system_status")
```

Comment:
I'm just now realizing that these additional bootloader vars need to be added to LK at some point before we can expect the feature to work on LK systems (right now I think that probably the bootloader vars would just silently get ignored when we write them, maybe we should change that too)

Reply:
Opened #9977 please take a look.